CMakeLists.txt

```diff
@@ -28,13 +28,11 @@ include(BuildOptionDefaults)
 # =============================================================================
 # Build options (boolean)
 # =============================================================================
-option(NRN_ENABLE_SHARED "Build shared libraries (otherwise static library)"
-       ${NRN_ENABLE_SHARED_DEFAULT})
+option(NRN_ENABLE_BINARY_SPECIAL "Build binary special"
+       ${NRN_ENABLE_BINARY_SPECIAL_DEFAULT})
```

Comment:
I'm a little confused about whether NRN_ENABLE_SHARED can be eliminated. I might be mistaken and need to look
into this in more detail but I thought  -DNRN_ENABLE_SHARED=OFF meant that libnrniv.a was constructed and nrniv
was a static executable. That did not mean that special was a binary. Even though nrniv was a static executable, it still dlopen'd libnrnmech.so. In autotools,  nrniv could be either statically or dynamically linked to libnrniv. In either
case it could dlopen libnrnmech.so either under user control or by a -dll arg or by the default of finding in the launch folder,  x86_64/.libs/libnrnmech.so.  The opposite of this in autotools was to build a full statically linked special
executable that linked statically against all the mod files and libnrniv.a. Or else (and I'm a bit shaky about this) build a special that linked against all the mod files and linked dynamically to libnrniv.so. All of these possiblities has their reasons at one historical time or another (eg. Blue Gene/L had no provision for dynamic linking (or dlopen) and so
even python with all necessary modules had to be linked statically into the special (not fun). Dynamic libnrniv.so is
needed for python, as in libnrnmech.so. But for nrniv, it would be sufficient to be fully static and launch in the parent
folder of x86_64/.libs/libnrnmech.so so that it comes in via dlopen. (By the way, the .libs folder is just an artifact of the
use of libtool to do the linking)

Reply:
AFAIK NRN_ENABLE_SHARED wasn't used anywhere, so we said to do some clean-up (didn't mean anything functional by that). 

@iomaganaris @pramodk  would you know anything about this? As far as I can tell, NRN_ENABLE_SHARED doesn't affect anything.

Reply:
You're right. It is unused. I think I just forgot to implement it in analogy to  the cmake for InterViews IV_ENABLE_SHARED. For the most part we prefer the latter as OFF so that libinterviews.a ends up inside  libnrniv.so. I think  that -DNRN_ENABLE_SHARED=OFF would mean that nrniv is a static executable (with respect to libnrniv.a)

Reply:
There has been mix of use cases about this topic. Let me try to summarise my understanding and how it should look like. Let's keep `special` out of the picture for now:

* `NRN_ENABLE_SHARED=OFF` : **Everything** should be built by static. I expect `x86_64/.libs/libnrnmech.so` also as static because user requested it. Now, I see that there is no need to build `libnrnmech.a` and all objects could be folded into `special` but lets say we keep the behaviour consistent.
* `NRN_ENABLE_SHARED=ON` : **Everything** should be built by shared. This would be default behaviour.

Depending upon this `NRN_ENABLE_SHARED`, `IV_ENABLE_SHARED` gets value ON or OFF. So I wouldn't remove those options.

Now about `special` : We need a flexibility to build special as real executable instead of just shell script. The shell script could be replaced by real binary in the future (IMO).
* NRN_ENABLE_BINARY_SPECIAL=OFF is default and in this case current behaviour is kept.
* NRN_ENABLE_BINARY_SPECIAL=ON, `special` will be real binary. It becomes shared / static binary based on value of `NRN_ENABLE_SHARED`.

I see that the fully static library is not that useful nowadays. But keeping that functionality in Make might be helpful in the future.

In BBP we want to do:   `NRN_ENABLE_SHARED=ON & NRN_ENABLE_BINARY_SPECIAL=ON` because end users expect ` `x86_64/.libs/libnrnmech.so` as shared library and we want to have `special` as real binary..

> so even python with all necessary modules had to be linked statically into the special (not fun). 

Oh bluegene! This was painful especially when we needed external Python modules! I don't see this use case any longer exist (anywhere). Python now days  is always shared library. If there will be only static library available, I wouldn't worry about this as this is mostly externally provided library.

Reply:
I see that `NRN_ENABLE_SHARED=OFF` conflicts with `dynamic loading` aspects (`NRN_ENABLE_MPI_DYNAMIC` or `NRN_ENABLE_PYTHON_DYNAMIC`). I wouldn't prioritise fully static build now as there are not sufficient use cases today.

@nrnhines : one of the main reason we want to have `special` as a binary is that with some MPI libraries (specially HPE-MPI which is in production on BBP system), the pinning of processes to cores is wrong! This results into ~25% slowdown in the execution of production simulations. We need a way to have `special` as a real binary and `libnrnmech` as shared library.

Reply:
@pramodk I think your understanding is pretty reasonable and I can adopt it for the most part.
I don't think that NRN_ENABLE_SHARED=OFF conflicts with  NRN_ENABLE_MPI_DYNAMIC=ON, etc. except that the name is a bit confusing as it does not indicate how "OFF" we mean.  The primary loss of functionality with libnrniv.a is that it cannot be loaded as a python module when python is launched (note that launch of nrniv
still allows use of python). At a minimum, NRN_ENABLE_SHARED=OFF means that the nrniv executable links against libnrniv.a. (And if IV_ENABLE_SHARED=OFF, then also links against libinterviews.a).  I presume it also means that it will prefer to link against any static library if there is both a .a and a .so file for the library. This could be useful for readline/ncurses binary distributions although it could be surprising if it links against the static python library libpython.a. although I think it will now do just that unless NRN_ENABLE_PYTHON_DYNAMIC=ON

Reply:
Hi @nrnhines , do you recall if for autotools having NRNMECH_DLL_STYLE=FALSE (`linux_nrnmech=no`) is compatible with RXD? This setup is currently failing for this PR in travis with cmake and we were wondering if it's actually a regression from autotools or not.